### PR TITLE
semaphore_osx: change mp_sem_timedwait to mp_time

### PR DESCRIPTION
**Pull request by kasper93**

### CODE REVIEW

**Review by Dudemanguy — APPROVED** [review]
LGTM if someone on macOS could double check that would be nice.

File: osdep/semaphore_osx.c [hunk 1]
```diff
@@ -67,32 +69,18 @@ int mp_sem_trywait(mp_sem_t *sem)
...
-                long long msec = (abs_timeout->tv_sec - tv.tv_sec) * 1000LL +
-                    abs_timeout->tv_nsec / 1000LL / 1000LL - tv.tv_usec / 1000LL;
-                if (msec > INT_MAX)
-                    msec = INT_MAX;
-                if (msec < 0)
-                    msec = 0;
-                timeout_ms = msec;
-            }
-        }
+        int64_t timeout = MPCLAMP(until - mp_time_ns(),
```
Comment (N-R-K) [inline osdep/semaphore_osx.c:78]: Ahem... I'd hoist out the function call to avoid double (or triple?) evaluation.

**Review by N-R-K — CHANGES_REQUESTED** [review]

File: osdep/semaphore_osx.c [hunk 2]
```diff
@@ -67,32 +69,18 @@ int mp_sem_trywait(mp_sem_t *sem)
...
+                                  MP_TIME_MS_TO_NS(INT_MIN),
+                                  MP_TIME_MS_TO_NS(INT_MAX));
+        timeout /= MP_TIME_MS_TO_NS(1);
```
Comment (N-R-K) [inline osdep/semaphore_osx.c:81]: Also, I think it'd be more accurate to do the clamping *after* the division.

**Review by N-R-K — CHANGES_REQUESTED** [review]

File: osdep/semaphore_osx.c (hunk 2 above)
Comment (N-R-K) [reply osdep/semaphore_osx.c:81]: But in any case, I think it's more natural to do it like this. And IMO more straight=forward to read as well.

```c
int64_t timeout = until - mp_time_ns();
timeout /= MP_TIME_MS_TO_NS(1);
timeout = MPCLAMP(timeout, INT_MIN, INT_MAX);
```

Maybe you can inline the division though.

**Comment by m154k1** [comment]
Works fine on macOS 14.0.

File: osdep/semaphore_osx.c (hunk 1 above)
Comment (kasper93) [reply osdep/semaphore_osx.c:78]: changed

File: osdep/semaphore_osx.c (hunk 2 above)
Comment (kasper93) [reply osdep/semaphore_osx.c:81]: changed, see now

**Review by N-R-K — APPROVED** [review]

**Pull request merged**